torch/_inductor/graph.py

```diff
@@ -1872,38 +1863,79 @@ def make_assert(expr: SympyBoolean, msg: str) -> None:
             f"new operations are:\n\n{format_new_defs()}"
         )
+        self.create_deferred_runtime_asserts(n, new_unbacked_defs)
+        return result
 
-        # Emit code for runtime asserts that can be inserted at this point.
-        for i0 in new_unbacked_defs:
-            ras = self.ras_by_symbol.pop(i0, [])
-            # NB: size-like not needed, we won't retrace
-            vr = shape_env.var_to_range[i0]
-            if not shape_env._default_unspecified_value_range().issubset(vr):
-
-                def is_convertible(s: Expr) -> bool:
-                    if s in (int_oo, -int_oo):
-                        return False
-                    try:
-                        int(s)
-                        return True
-                    except TypeError:
-                        return False
-
-                if is_convertible(vr.lower):
-                    make_assert(i0 >= vr.lower, f"{i0} >= {vr.lower}")
-                if is_convertible(vr.upper):
-                    make_assert(i0 <= vr.upper, f"{i0} <= {vr.upper}")
-
-            for ra in ras:
-                fvs = free_unbacked_symbols(ra.expr)
-                missing = fvs - self.bound_unbacked_symbols
-                if missing:
-                    i1 = min(missing, key=str)
-                    self.ras_by_symbol.setdefault(i1, []).append(ra)
-                else:
-                    make_assert(ra.expr, f"{ra.expr}")
+    def create_deferred_runtime_asserts(
+        self, n: torch.fx.Node, new_unbacked_defs: OrderedSet[sympy.Symbol]
+    ) -> None:
+        # [NOTE] Codegen runtime asserts in Inductor
+        # In TorchInductor, we need to generate runtime asserts directly, but
+        # for AOTInductor, we can re-use the asserts from exported graphs.
+        #
+        # For export, our strategy for generating runtime asserts was that
+        # Dynamo or non-strict export would insert them into the FX graph
+        # after finishing tracing, and we would attempt to code generate
+        # them based on the FX graph. This is a good strategy for export,
+        # where we immediately export the graph. However, this strategy
+        # was afflicted by problems in eager, where we reuse the same
+        # ShapeEnv as before. In particular, on subsequent graph passes,
+        # we would immediately turn all of these assertions into noops,
+        # because when we evaluated their expressions, we would see that
+        # because we had a deferred runtime assert in the ShapeEnv, we
+        # know "oh, of course this expression is True" already.
 
-        return result
+        def make_assert(expr: SympyBoolean, msg: str) -> None:
+            assert_op = ir.AssertScalar(expr, msg)
+            self.register_buffer(assert_op, set_name=True)
+            self.register_operation(assert_op)
+
+        if (
+            full_aoti_runtime_assert()
+            and n.target == torch.ops.aten._assert_scalar.default
+            and self.aot_mode
```

Comment:
in general probably eventually we want to add those even for non aot mode.
regular torch.compile should emit those, apparently we can have them if prefer_deferred_runtime_asserts_over_guards.
I am not sure how urgnet or important that is right now.